daemon/cmd/ipam.go

```diff
@@ -192,38 +197,82 @@ func coalesceCIDRs(rCIDRs []string) (result []string) {
 	return
 }
 
-func (d *Daemon) allocateDatapathIPs(family types.NodeAddressingFamily) (routerIP net.IP, err error) {
-	// Avoid allocating external IP
-	d.ipam.ExcludeIP(family.PrimaryExternal(), "node-ip", ipam.PoolDefault())
+type ipamAllocateIP interface {
+	AllocateIPWithoutSyncUpstream(ip net.IP, owner string, pool ipam.Pool) (*ipam.AllocationResult, error)
+}
 
-	// (Re-)allocate the router IP. If not possible, allocate a fresh IP.
-	// In that case, removal and re-creation of the cilium_host is
-	// required. It will also cause disruption of networking until all
-	// endpoints have been regenerated.
-	var result *ipam.AllocationResult
-	routerIP = family.Router()
-	if routerIP != nil {
-		result, err = d.ipam.AllocateIPWithoutSyncUpstream(routerIP, "router", ipam.PoolDefault())
+// reallocateDatapathIPs attempts to reallocate the old router IP from IPAM.
+// It prefers fromFS over fromK8s. If neither IPs can be re-allocated, log
+// messages are emitted and the function returns nil.
+func reallocateDatapathIPs(alloc ipamAllocateIP, fromK8s, fromFS net.IP) (result *ipam.AllocationResult) {
+	if fromK8s == nil && fromFS == nil {
+		// We do nothing in this case because there are no router IPs to restore.
+		return nil
+	}
+
+	// If we have both an IP from the filesystem and an IP from the Kubernetes
+	// resource, and they are not equal, emit a warning.
+	if fromK8s != nil && fromFS != nil && !fromK8s.Equal(fromFS) {
+		log.Warnf(
+			mismatchRouterIPsMsg,
+			fromK8s, fromFS, option.LocalRouterIPv4, option.LocalRouterIPv6,
+		)
+		// Above is just a warning; we still want to set the router IP regardless.
+	}
+
+	// Router IPs from the filesystem are preferred over the IPs found
+	// in the Kubernetes resource (Node or CiliumNode), because we consider the
+	// filesystem to be the most up-to-date source of truth.
+	var err error
+	if fromFS != nil {
+		result, err = alloc.AllocateIPWithoutSyncUpstream(fromFS, "router", ipam.PoolDefault())
 		if err != nil {
-			log.Warn("Router IP could not be re-allocated. Need to re-allocate. This will cause brief network disruption")
-
-			// The restored router IP is not part of the allocation range.
-			// This indicates that the allocation range has changed.
-			deleteHostDevice()
+			log.WithError(err).Warnf(
+				"Unable to restore router IP %s from filesystem",
+				fromFS,
+			)
```

Comment:
Let's move this to a structured log for the IP address rather than a dynamically formatted string